src/commissioner/Discoverer.h

```diff
@@ -0,0 +1,134 @@
+
+struct Delegate
+{
+    virtual ~Delegate()                 = default;
+    virtual void OnDiscovery()          = 0;
+    virtual void OnDiscovererShutdown() = 0;
+};
+
+struct ShutdownToken
+{
+    ShutdownToken(Delegate ** delegate) : mDelegate(delegate) {}
+    ~ShutdownToken() { (*mDelegate)->OnDiscovererShutdown(); }
+
+private:
+    Delegate ** mDelegate;
+};
+
+struct Joinable
+{
+    Joinable(Delegate ** delegate) : mDelegate(delegate) {}
+
+protected:
+    bool InProgress() { return mShutdownToken.get() != nullptr ? true : false; }
+    void RetainShutdownToken(Platform::SharedPtr<ShutdownToken> token) { mShutdownToken = token; }
+    void ReleaseShutdownToken() { mShutdownToken = std::shared_ptr<ShutdownToken>(nullptr); }
+    Platform::SharedPtr<ShutdownToken> GetShutdownToken() { return mShutdownToken; }
+
+    Delegate ** mDelegate;
+
+private:
+    Platform::SharedPtr<ShutdownToken> mShutdownToken;
+};
+
+#if CONFIG_NETWORK_LAYER_BLE
+class BleDiscoverer : Joinable
+{
+public:
+    BleDiscoverer(Controller::DeviceControllerSystemState & systemState, Delegate ** delegate);
+    ~BleDiscoverer();
+
+    void Shutdown();
+    CHIP_ERROR StartBleDiscovery(SetupPayload & payload, Platform::SharedPtr<ShutdownToken> token);
+    CHIP_ERROR GetNextBleCandidate(Commissionee & commissionee);
+    void TallyBleDiscovery(BLE_CONNECTION_OBJECT connection);
+    void TallyBleFailure();
+
+private:
+    static void OnBleDiscoverySuccess(void * context, BLE_CONNECTION_OBJECT connection);
+    static void OnBleDiscoveryError(void * context, CHIP_ERROR err);
+    Controller::DeviceControllerSystemState & mSystemState;
+    chip::Optional<BLE_CONNECTION_OBJECT> mBleConnection;
+};
+#endif // CONFIG_NETWORK_LAYER_BLE
+
+#if CHIP_DEVICE_CONFIG_ENABLE_DNSSD
+class DnssdDiscoverer : Controller::AbstractDnssdDiscoveryController, Controller::DeviceDiscoveryDelegate, Joinable
+{
+public:
+    DnssdDiscoverer(Controller::DeviceControllerSystemState & systemState, Delegate ** delegate);
+    ~DnssdDiscoverer();
+
+    void Shutdown();
+    CHIP_ERROR StartDnssdDiscovery(SetupPayload & payload, Platform::SharedPtr<ShutdownToken> token);
+    CHIP_ERROR GetNextDnssdCandidate(Commissionee & commissionee);
+
+private:
+    DiscoveredNodeList GetDiscoveredNodes() { return DiscoveredNodeList(mNodes); }
+    void OnDiscoveredDevice(const chip::Dnssd::DiscoveredNodeData & node) { (*mDelegate)->OnDiscovery(); }
+    void OnNodeIdResolved(const Dnssd::ResolvedNodeData & nodeData) {}
+    void OnNodeIdResolutionFailed(const PeerId & peerId, CHIP_ERROR error) {}
+
+    CHIP_ERROR GetValidRecord();
+    CHIP_ERROR GetAddress(chip::Optional<Transport::PeerAddress> & address,
+                          chip::Optional<ReliableMessageProtocolConfig> & mrpConfig);
+
+    Controller::DeviceControllerSystemState & mSystemState;
+    Dnssd::DiscoveredNodeData mNodes[CHIP_DEVICE_CONFIG_MAX_DISCOVERED_NODES];
+    size_t mNodeIdx = 0;
+    int mIpIdx      = 0;
```

Comment:
A brief doxygen `///<` style comment here explaining the purpose of these instance variables would be helpful.

Reply:
Will do.